scripts/fetch_service_refs.py

```diff
@@ -0,0 +1,96 @@
+import argparse
+import json
+import os
+import sys
+import time
+import urllib.error
+import urllib.request
+from typing import Any, Dict, Optional
+
+GO_SERVER_URL = os.environ.get(
+    "GO_SERVER_URL", "http://gocd-server.gocd.svc.cluster.local:8153/go"
+)
+
+
+def pipeline_passed(pipeline: Dict[str, Any]) -> bool:
+    # stage["result"] isn't populated if it isn't run
+    # the other possible statuses are Unknown (not run yet), Cancelled, Failed
+    return all(stage["status"] == "Passed" for stage in pipeline["stages"])
+
+
+# print the most recent passing sha for a repo
+def main(pipeline_name: str = "deploy-snuba", repo: str = "snuba") -> int:
+    GOCD_ACCESS_TOKEN = os.environ.get("GOCD_ACCESS_TOKEN")
+    if not GOCD_ACCESS_TOKEN:
+        raise SystemExit(
+            """
+            GOCD_ACCESS_TOKEN not set. It should be an access token belonging to bot@sentry.io.
+            """
+        )
+    fetch_url: Optional[str] = f"{GO_SERVER_URL}/api/pipelines/{pipeline_name}/history"
+
+    while fetch_url:
+        req = urllib.request.Request(
+            fetch_url,
+            headers={
+                "Accept": "application/vnd.go.cd.v1+json",
+                "Authorization": f"bearer {GOCD_ACCESS_TOKEN}",
+            },
+        )
+        try:
+            resp = urllib.request.urlopen(req)
+        except urllib.error.HTTPError as e:
+            raise SystemExit(f"Failed to fetch pipeline history:\n{e.read().decode()}")
+
+        print(
+            "fetching pipeline history for", pipeline_name, fetch_url, file=sys.stderr
+        )
+        data = json.loads(resp.read())
+
+        if "_links" in data and "next" in data["_links"]:
+            fetch_url = data["_links"]["next"]["href"]
+        else:
+            fetch_url = None
+        rev = None
+
+        for pipeline in sorted(
+            data["pipelines"], key=lambda _: int(_["counter"]), reverse=True
+        ):
+            # Look at the most recent passing pipeline,
+            # and get its deployment revision for the main material.
+            if pipeline_passed(pipeline):
+
+                for r in pipeline["build_cause"]["material_revisions"]:
+                    # example material description format... `in` is good enough
+                    # 'URL: git@github.com:getsentry/devinfra-example-service.git, Branch: main'
+                    if (
+                        f"git@github.com:getsentry/{repo}.git"
+                        in r["material"]["description"]
+                    ):
+                        rev = r["modifications"][0]["revision"]
+                        print(rev)
+                        return 0
+                    elif (
+                        f"https://github.com/getsentry/{repo}.git"
+                        in r["material"]["description"]
+                    ):
+                        rev = r["modifications"][0]["revision"]
+                        print(rev)
+                        return 0
+        time.sleep(1)
```

Comment:
Are there cases where you might never exit this while loop? Should we have a maximum number of retries or maximum time for the job to check?

Reply:
we should always exist since we fetch the pagination link for the response and break if it is null. I added a max fetches parameter anyway just in case.